### PR TITLE
GH-15050: [Java][Docs] Update and consolidate Memory documentation

### LINKED ISSUE #15050: [Java] [Docs] Reorganize and update Java tutorial Memory page
### Describe the enhancement requested

Move selected text from the package JavaDoc to the rst/html documentation to avoid duplication and maintenance issues. Update docs to include mention of simplified handling of C++ memory allocations. 

### Component(s)

Documentation, Java

### PR DESCRIPTION
**Pull request by lwhite1**
* Closes: #15050

### CODE REVIEW

**Comment by lwhite1** [comment]
@lidavidm Would you take a look when you have time? Thanks very much.

File: docs/source/java/memory.rst [hunk 1]
```diff
@@ -284,3 +302,159 @@ Finally, enabling the ``TRACE`` logging level will automatically provide this st
...
+   |
+   |-+ BufferLedger 1 ==> Allocator A (owning)
+   | ` - ArrowBuf 1
+   |-+ BufferLedger 2 ==> Allocator B (non-owning)
+   | ` - ArrowBuf 2
+   |-+ BufferLedger 3 ==> Allocator C (non-owning)
+     | - ArrowBuf 3
+     | - ArrowBuf 4
+     ` - ArrowBuf 5
+   </pre>
```
Comment (lidavidm) [inline docs/source/java/memory.rst:412]: nit: I think you can use `.. code-block:: none` and remove the `<pre>` tags. (Not that I expect anyone to render the docs to anything but HTML…)

**Review by lidavidm — APPROVED** [review]
This is great. Thanks for reworking the docs!

**Comment by lwhite1** [comment]
ok. let me give that a try

**Comment by lwhite1** [comment]
I made the formatting fix so this should be set when the tests complete. Thank you for your review.

**Pull request merged**